eth/executionclient/execution_client.go

```diff
@@ -234,6 +227,90 @@ func (ec *ExecutionClient) fetchLogsInBatches(ctx context.Context, startBlock, e
 	return logCh, errCh
 }
 
+// subdivideLogFetch handles log fetching with automatic subdivision on rate limit errors.
+// It first attempts a direct fetch, and if that fails with a rate limit error, it subdivides
+// the block range and tries again with smaller chunks.
+func (ec *ExecutionClient) subdivideLogFetch(ctx context.Context, q ethereum.FilterQuery) ([]ethtypes.Log, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-ec.closed:
+		return nil, ErrClosed
+	default:
+	}
+
+	logs, err := ec.client.FilterLogs(ctx, q)
+	if err == nil {
+		return logs, nil
+	}
+
+	if !isRPCRateLimitError(err) || q.FromBlock == nil || q.ToBlock == nil {
+		return nil, err
+	}
+
+	fromBlock := q.FromBlock.Uint64()
+	toBlock := q.ToBlock.Uint64()
+
+	// check if we have at least 2 blocks
+	if fromBlock >= toBlock {
+		return nil, err
+	}
+
+	ec.logger.Info("subdividing log fetch due to rate limit",
+		fields.FromBlock(fromBlock),
+		fields.ToBlock(toBlock))
+
+	midBlock := fromBlock + (toBlock-fromBlock)/2
+
+	firstHalfQuery := q
+	firstHalfQuery.FromBlock = new(big.Int).SetUint64(fromBlock)
+	firstHalfQuery.ToBlock = new(big.Int).SetUint64(midBlock)
+
+	secondHalfQuery := q
+	secondHalfQuery.FromBlock = new(big.Int).SetUint64(midBlock + 1)
+	secondHalfQuery.ToBlock = new(big.Int).SetUint64(toBlock)
+
+	type fetchResult struct {
+		logs []ethtypes.Log
+		err  error
+	}
+
+	firstHalfCh := make(chan fetchResult, 1)
+	secondHalfCh := make(chan fetchResult, 1)
+
+	go func() {
+		fetchedLogs, fetchErr := ec.subdivideLogFetch(ctx, firstHalfQuery)
+		firstHalfCh <- fetchResult{fetchedLogs, fetchErr}
+	}()
+
+	go func() {
+		fetchedLogs, fetchErr := ec.subdivideLogFetch(ctx, secondHalfQuery)
+		secondHalfCh <- fetchResult{fetchedLogs, fetchErr}
+	}()
+
+	firstResult := <-firstHalfCh
+	if firstResult.err != nil {
+		return nil, firstResult.err
+	}
+
+	secondResult := <-secondHalfCh
+	if secondResult.err != nil {
+		return nil, secondResult.err
+	}
+
+	totalLogs := len(firstResult.logs) + len(secondResult.logs)
+	combinedLogs := make([]ethtypes.Log, 0, totalLogs)
+	combinedLogs = append(combinedLogs, firstResult.logs...)
+	combinedLogs = append(combinedLogs, secondResult.logs...)
```

Comment:
imo we shouldn't call concurrently

yes, it's (might be) faster, but not worth the added complexity and load on the EL

especially given the reason we're dividing in the first place is because this specific range of blocks has tons of events and is heavy to query :)

Reply:
reverted, I struggled with my thoughts not to add this, but I couldnt stop myself 😅